### PR TITLE
Introduce auto-compaction policies to Controller

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR introduces to `Controller` an enum `ReadPolicy`, which allows the coordinator to indicate a policy the controller should use to automatically advance the read capability. There are two variants, `ValidFrom(frontier)` and `LagWriteFrontier(fn)`. The former fixes the read capability at `frontier`, and the latter applies `fn` each time the write frontier (`upper`) updates.

The current calls to `allow_compaction()` are translated into `ValidFrom` policies, which allow compaction up to the indicated frontiers. However, this may eventually be deprecated in favor of the new `set_read_policies` which allows you to do that, or to use the `LagWriteFrontier` variant to allow autocompaction, even while you are not looking.

None of this has been tested yet, as it needs the coordinator to opt in to it. So, this is mostly a design, and we'll see the bugs once we point the coordinator at it in anger.

It is not presently as ergonomic to return to `ValidFrom` policies, as I haven't sorted out how one atomically grabs the current frontier into the policy (vs reading it, and then attempting to set the policy, which may become invalid immediately). Potentially, `ValidFrom([0])` would stop all compaction, not at that frontier of course, at which point compaction would be locked down. So, questions exist.

This PR only manages Compute collection compaction. The logic would be identical for sources, but sources do not generally produce write frontier progress statements (they produce the conditional upper things, which .. probably do not justify compacting the data until the updates are made durable? I'm hoping to finesse this by having us remove the coordinator as the source of durability.

### Motivation

The coordinator would like to be out of the business of maintaining compaction of collections. 

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by frankmcsherry** [comment]
In principle this is ready for review. It is passing, and I'm not certain what it should do next. There are certainly open questions about e.g. coordinator ergonomics attaching to this, and it is possible this is unusable in its current form. However, I also think it is harmless, and we could next attempt to pivot over to having the coordinator try it, and see what pain points it has.

File: src/dataflow-types/src/client/controller/compute.rs [hunk 1]
```diff
@@ -312,32 +314,45 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeController<'a, C, T> {
 
         // // Validate that the ids exist.
         // self.validate_ids(frontiers.iter().map(|(id, _)| *id))?;
+        let policies = frontiers
```
Comment (benesch) [inline src/dataflow-types/src/client/controller/compute.rs:317]: I know this is existing code, but is there a reason that the ID validation one line up is disabled?

File: src/dataflow-types/src/client/controller/compute.rs [hunk 2]
```diff
@@ -312,32 +314,45 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeController<'a, C, T> {
...
+            .into_iter()
+            .map(|(id, frontier)| (id, ReadPolicy::ValidFrom(frontier)));
+        self.set_read_policy(policies.collect()).await;
+    }
 
-        let mut updates = BTreeMap::new();
-        for (id, mut frontier) in frontiers.into_iter() {
-            // If-let to evade some identifiers incorrectly sent to us.
+    pub async fn set_read_policy(&mut self, policies: Vec<(GlobalId, ReadPolicy<T>)>) {
```
Comment (benesch) [inline src/dataflow-types/src/client/controller/compute.rs:323]: Mind adding a doc string on this one? Wasn't clear to me whether this wholesale replaced read policies, or just overwrote the read policies for the specified IDs.

**Review by benesch — APPROVED** [review]
Seems good. Only one way to find out whether this works for the coordinator, and that's to wire it up! Very much on board with merging as is.

File: src/dataflow-types/src/client/controller/compute.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/compute.rs:317]: It crashes on `main` when it is enabled (until https://github.com/MaterializeInc/materialize/pull/10835 lands).

File: src/dataflow-types/src/client/controller/compute.rs (hunk 2 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/compute.rs:323]: Will do.

File: src/dataflow-types/src/client/controller.rs [hunk 3]
```diff
@@ -131,3 +128,15 @@ impl<C> Controller<C> {
         }
     }
 }
+
+/// Compaction policies for collections maintained by `Controller`.
+pub enum ReadPolicy<T = mz_repr::Timestamp> {
```
Comment (mjibson) [inline src/dataflow-types/src/client/controller.rs:133]: I think you can leave out `= mz_repr::Timestamp` and nothing complains?

File: src/dataflow-types/src/client/controller.rs [hunk 4]
```diff
@@ -131,3 +128,15 @@ impl<C> Controller<C> {
...
+    /// Maintain the collection as valid from this frontier onward.
+    ValidFrom(Antichain<T>),
+    /// Maintain the collection as valid from a function of the write frontier.
+    ///
+    /// This function will only be re-evaluated when the write frontier changes.
+    /// If the intended behavior is to change in response to external signals,
+    /// consider using the `StopAt` variant to manually pilot compaction.
```
Comment (mjibson) [inline src/dataflow-types/src/client/controller.rs:140]: Did `StopAt` get renamed to `ValidFrom`?

File: src/dataflow-types/src/client/controller/compute.rs [hunk 5]
```diff
@@ -361,6 +376,41 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeController<'a, C, T> {
         Ok(())
     }
 
+    /// Accept write frontier updates from the compute layer
```
Comment (mjibson) [inline src/dataflow-types/src/client/controller/compute.rs:387]: missing period at EOS

**Review by mjibson — APPROVED** [review]

File: src/dataflow-types/src/client/controller.rs (hunk 4 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller.rs:140]: Oh dear. It did.

File: src/dataflow-types/src/client/controller.rs (hunk 3 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller.rs:133]: You are right, and it even would have saved me one round of errors where I forgot the argument and it defaulted to this, despite needing to be generic.

File: src/dataflow-types/src/client/controller/compute.rs (hunk 5 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/compute.rs:387]: fixed!

**Pull request merged**